File: app/src/main/java/com/iamagamedev/mybrandnewgame/collisions/SpellCollisions.java
package com.iamagamedev.mybrandnewgame.collisions;

import android.os.Handler;
import android.os.Looper;

import com.iamagamedev.mybrandnewgame.Constants.CharConstants;
import com.iamagamedev.mybrandnewgame.LevelManager;
import com.iamagamedev.mybrandnewgame.SoundManager;
import com.iamagamedev.mybrandnewgame.gameObjects.GameObject;
import com.iamagamedev.mybrandnewgame.gameObjects.spells.SpellObject;

public class SpellCollisions {

    private static Handler handler = new Handler(Looper.getMainLooper());

    public static void checkForSpellCollisions(GameObject go, final LevelManager lm) {
        if (go == lm.spellObject) return;
        int col = 0;
        col = lm.spellObject.checkCollisions(go.getRectHitBoxRight(), go.getRectHitBoxLeft(),
                go.getRectHitBoxTop(), go.getRectHitBoxBottom());
        if (col > 0) {
            switch (go.getType()) {
                case CharConstants.ENEMY_SHIELD:
                    spellIsUsed(lm.spellObject);
                    break;
                case CharConstants.WALL:
                    go.setActive(false);
                    spellIsUsed(lm.spellObject);
                    break;
                case CharConstants.ENEMY:
                case CharConstants.ENEMY_TANKU:
                    break;
                default:
                    break;
            }
        }
    }

    private static void spellIsUsed(final SpellObject spellObject) {
        if (!spellObject.isActive()) return;
        SoundManager.getInstance().playSound("explode");
        spellObject.setActive(false);
        spellObject.setMoves(false);
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                spellObject.setWorldLocation(-1000, -1000, 0);
            }
        }, 1000);
    }
}

File: app/src/main/java/com/iamagamedev/mybrandnewgame/collisions/SpellCollisions.kt
package com.iamagamedev.mybrandnewgame.collisions

import android.os.Handler
import android.os.Looper
import com.iamagamedev.mybrandnewgame.constants.CharConstants
import com.iamagamedev.mybrandnewgame.LevelManager
import com.iamagamedev.mybrandnewgame.SoundManager
import com.iamagamedev.mybrandnewgame.gameObjects.EnemyObject
import com.iamagamedev.mybrandnewgame.gameObjects.GameObject

object SpellCollisions {
    private val handler = Handler(Looper.getMainLooper())
    fun checkForSpellCollisions(go: GameObject) {
        if (LevelManager.spellObject == null || go === LevelManager.spellObject || !LevelManager.spellObject!!.isActive) return
        var col = 0
        col = LevelManager.spellObject!!.checkCollisions(go.rectHitBoxRight, go.rectHitBoxLeft,
                go.rectHitBoxTop, go.rectHitBoxBottom)
        if (col > 0) {
            when (go.type) {
                CharConstants.ENEMY, CharConstants.ENEMY_TANKU -> {
                    spellIsUsed()
                    if (!(go as EnemyObject).shieldIsActive) {
                        go.health = go.health - LevelManager.spellObject!!.damage
                        if (go.health <= 0) {
                            go.isActive = false
                            go.isDead = true
                        }
                    }
                }
                CharConstants.WALL -> {
                    go.isActive = false
                    spellIsUsed()
                }
                CharConstants.ENEMY_SHIELD -> {
                    spellIsUsed()
                }
                else -> {
                }
            }
        }
    }

    private fun spellIsUsed() {
        if (!LevelManager.spellObject!!.isActive) return
        SoundManager.instance?.playSound("explode")
        LevelManager.spellObject?.isActive = false
        LevelManager.spellObject?.isMoves = false
        handler.postDelayed({
            LevelManager.spellObject?.updatePosition((-100).toFloat(), (-100).toFloat())
            LevelManager.spellObject?.isVisible = false
        }, 200)
    }
}